core/sys/etimer.c

```diff
@@ -150,27 +156,44 @@ etimer_request_poll(void)
 static void
 add_timer(struct etimer *timer)
 {
-  struct etimer *t;
+  clock_time_t now = clock_time();
+  struct process *proc = PROCESS_CURRENT();
+  struct etimer *t_this;
+  struct etimer *t_last;
+
+  if(proc == PROCESS_NONE) {
+    /* don't add an etimer with no process */
+    PRINTF("etimer: proc is NONE\n");
+    return;
+  }
 
   etimer_request_poll();
 
   if(timer->p != PROCESS_NONE) {
-    for(t = timerlist; t != NULL; t = t->next) {
-      if(t == timer) {
-	/* Timer already on list, bail out. */
-        timer->p = PROCESS_CURRENT();
-	update_time();
-	return;
-      }
-    }
+    etimer_stop(timer);
   }
 
-  /* Timer not on list. */
-  timer->p = PROCESS_CURRENT();
-  timer->next = timerlist;
-  timerlist = timer;
+  timer->p = proc;
 
-  update_time();
+  if(timerlist == NULL || etimer_lte(timer, timerlist, now)) {
+    timer->next = timerlist;
+    timerlist = timer;
+    update_next_expiration();
+    return;
+  }
+
+  t_last = timerlist;
+  for(t_this = timerlist->next; t_this != NULL; t_this = t_this->next) {
+    if(etimer_gte(timer, t_last, now) && etimer_lte(timer, t_this, now)) {
+      timer->next = t_this;
+      t_last->next = timer;
+      return;
+    }
+    t_last = t_this;
+  }
+
+  t_last->next = timer;
+  timer->next = NULL;
```

Comment:
The list is always sorted; and, it is searched linearly.  Therefore, the first timer test is redundant; it stays true until after the second test has found the insertion point.  Also, appending to the list is the same as inserting in front of the list's `NULL`.

The above 10 lines can be replaced by this code:

```C
    if(etimer_lte(timer, t_this, now)) {
      break;
    }
    t_last = t_this;
  }
  t_last->next = timer;
  timer->next = t_this;
```